File: backend/src/models/bio_data.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParsedHeaderInfo {
    pub primary_id: String,
    pub kingdom: Option<String>,
    pub phylum: Option<String>,
    pub class: Option<String>,
    pub order: Option<String>,
    pub family: Option<String>,
    pub genus: Option<String>,
    pub species: Option<String>,
    pub other_info: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FastaSequence {
    #[serde(flatten)]
    pub header_info: ParsedHeaderInfo,
    pub sequence: String, 
}

#[derive(Debug, Deserialize)]
pub struct KmerSearchRequest {
    pub query_sequence: String,
    pub k: usize,               
}

#[derive(Debug, Clone, Serialize)]
pub struct SimilarityResult {
    #[serde(flatten)] 
    pub header_info: ParsedHeaderInfo,
    pub sequence: String,
    pub similarity_score: f64,
}

#[derive(Debug, Serialize)]
pub struct KmerSearchResponse {
    pub results: Vec<SimilarityResult>,
}

File: backend/src/models/bio_data.py
from pydantic import BaseModel
from typing import Optional, List

class ParsedHeaderInfo(BaseModel):
    primary_id: str
    kingdom: Optional[str] = None
    phylum: Optional[str] = None
    class_: Optional[str] = None
    order: Optional[str] = None
    family: Optional[str] = None
    genus: Optional[str] = None
    species: Optional[str] = None
    other_info: Optional[str] = None

    class Config:
        # Allow 'class' as a field name
        allow_population_by_field_name = True
        fields = {"class_": "class"}

class FastaSequence(BaseModel):
    header_info: ParsedHeaderInfo
    sequence: str

class KmerSearchRequest(BaseModel):
    query_sequence: str
    k: int

class SimilarityResult(BaseModel):
    header_info: ParsedHeaderInfo
    sequence: str
    similarity_score: float

class KmerSearchResponse(BaseModel):
    results: List[SimilarityResult]
